### PR TITLE
Server-side: Implement Endpoints, Mongodb Setup, Testing Setup

### PR DESCRIPTION
**Pull request by Sherry2001**
PR Includes
- endpoints for items and categories
- mongo db connection set up and Mongoose schemas
- set up for testing and sample tests 

TODO: write tests for item and category routers. Will be in another PR

### CODE REVIEW

**Comment by Sherry2001** [comment]
Commits e019f2c till 46c4c20 are changes to itemRouters.js for @cynthia-doc to check (tried switching to await/more compact)

Commits 0b52e71 b756653 02635c9 7fec970 are for comments and code from PR #8 @Anthony-Fiddes

File: server/api/categoryRouter.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
+/**
+ * Router for item actions
+ */
+const Category = require('../db/models/category');
+const express = require('express');
+const Item = require('../db/models/item');
+const router = express();
+router.use(express.json());
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:8]: This is required for tests on categoryRouter to work

File: server/api/categoryRouter.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Create a new category
+ * 
+ * req.body: {title: String}
+ * 
+ * response: {success: boolean, message: String}
+ */
+router.post('/createCategory', (req, res, next) => {
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:17]: next must be passed to these functions, as it refers to the callback function. There is a reference error if it's not included

File: server/api/categoryRouter.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ * 
+ * response: [{title: String, 
+ *             items: [{
+ *               item objects from db 
+ *             }]
+ *            }]
+ */
+router.get('/getCategoryItems', async (req, res, next) => {
+  const categoryId = req.body.categoryId;
+  let response = {};
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:68]: This code should also go into the try block

File: server/api/categoryRouter.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  try {
+    //get category name and a list of item ids
+    const categoryData = await Category.findOne({ _id: categoryId }, 'title items', (error, data)).exec();
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:71]: This will error, model it after getCategories and add tests (error and data are undefined variables)

File: server/api/categoryRouter.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ * req.body : {categoryId: mongoose.objectId}
+ * 
+ * response: {
+ *            success: boolean,
+ *            message: String,
+ *           }
+ * error 
+ */
+router.get('/deleteCategory', async (req, res, next) => {
+  const categoryId = req.body.categoryId;
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:100]: This should go in the try block because it can error

File: server/api/categoryRouter.js (hunk 1 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:8]: ohh ok!

File: server/api/categoryRouter.js (hunk 2 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:17]: oh okay! will add it to item router too then

File: server/api/categoryRouter.js (hunk 4 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:71]: yes good catch, forgot to remove!

File: server/api/categoryRouter.js (hunk 3 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:68]: okay

File: server/api/categoryRouter.js [hunk 6]
```diff
@@ -36,14 +37,18 @@ router.post('/createCategory', (req, res) => {
...
-  const userId = req.body.userId;
+router.get('/getCategories', async (req, res, next) => {
   try {
-    const categoriesData = await Category.find({ userId: userId }, 'title', (error, data)).exec(); 
+    const userId = req.body.userId;
+    const categoriesData = await Category.find({ userId: userId }, 'title', function (err, data) {
+      if (err) {
+        next(err);
+      }
+    });
```
Comment (Sherry2001) [inline server/api/categoryRouter.js]: My bad for not deleting (error, data) on the original line 42, will do that. 

But for lines 44-46: I thought using await allows you to not do error handling through the callback, and the try block will just catch it. [https://stackoverflow.com/questions/48601249/how-to-return-mongoose-query-results-without-callback](url)

Comment (Sherry2001) [reply server/api/categoryRouter.js]: so technically the callback and error handling in it shouldn't be needed?

Comment (Anthony-Fiddes) [reply server/api/categoryRouter.js]: Yes, you're correct! Awesome catch

File: server/api/categoryRouter.js (hunk 1 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js]: oh actually, if you require index.js and use the full url for categories which is '/category/endpointName' it works, so we don't have to do router.use for each router. 
6bff5ca

File: server/api/itemRouter.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
+const Item = require('../db/models/item');
+const Category = require('../db/models/category');
+const express = require('express');
+
```
Comment (Anthony-Fiddes) [inline server/api/itemRouter.js:4]: Add todo to add testing

File: server/api/categoryRouter.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
+/**
+ * Router for item actions
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:2]: Should this be "Router for category actions?"

File: server/test/categoryRouter.test.js [hunk 9]
```diff
@@ -1,5 +1,6 @@
 const Category = require('../db/models/category');
-const categoryRouter = require('../api/categoryRouter');
+// const categoryRouter = require('../api/categoryRouter');
+const server = require('../index');
```
Comment (Sherry2001) [inline server/test/categoryRouter.test.js:3]: @Anthony-Fiddes

File: server/api/categoryRouter.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+ */
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:3]: Add a todo to add testing

Comment (Anthony-Fiddes) [reply server/api/categoryRouter.js:3]: Should these be official github issues?

File: server/test/categoryRouter.test.js (hunk 9 above)
Comment (Sherry2001) [reply server/test/categoryRouter.test.js:3]: this allows us to not do router.use(express.json) every time

File: server/api/categoryRouter.js (hunk 10 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:3]: sure, about github issues- we could? but the TODOs for testing would be in the category test file tho right?

File: server/api/categoryRouter.js (hunk 8 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:2]: yes!!

File: server/api/itemRouter.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+ * req.body: {itemId: mongoose.objectId,
+ *            categoryId: mongoose.objectId}
+ * 
+ * response: {success: boolean,
+ *            message: String}
+ */
+router.put('/addItemToCategory', async (req, res) => {
+  const itemId = req.body.itemId;
+  const categoryId = req.body.categoryId;
+  let response = {};
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/itemRouter.js:70]: yep, will apply all prev changes to itemRouter.js!

File: server/api/itemRouter.js (hunk 7 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:4]: okie

File: server/api/categoryRouter.js (hunk 2 above)
Comment (cynthia-doc) [reply server/api/categoryRouter.js:16]: Just added all of them:)))

File: server/test/categoryRouter.test.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+});
+
+describe('categoryRouter', function () {
+  describe('/getCategories', function () {
+    it('should return an empty list when no userId is provided in the request', async function () {
+      chai
+        .request(server)
+        .get('/category/getCategories')
+        .end(function (err, res) {
+          expect(err).to.be.null;
```
Comment (Sherry2001) [inline server/test/categoryRouter.test.js:51]: @Anthony-Fiddes can the handleErr(err) helper method you wrote be able to be used here? same in the next test

File: server/api/categoryRouter.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+      response.successs = false;
+      response.message = 'failed to add category';
+      next(err);
+    } else {
+      response.success = true;
+      response.message = 'created new category';
+    }
+    res.json(response);
+  });
+});
```
Comment (Anthony-Fiddes) [inline server/api/categoryRouter.js:30]: Should this have some sort of error catching around it?

File: server/api/itemRouter.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+ *            highlight: String,
+ *            userId: mongoose.ObjectId,
+ *            placesId: id(optional),
+ *            comment: String (optional),}
+ * 
+ * response: {success: boolean,
+ *            message: String}
+ */
+router.post('/addItem', (req, res, next) => {
+  const newItem = req.body;
```
Comment (Anthony-Fiddes) [inline server/api/itemRouter.js:19]: Same comment as above, can this statement error, and should it be handled?

File: server/test/categoryRouter.test.js (hunk 9 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js:2]: I love her!

File: server/api/itemRouter.js (hunk 14 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:19]: I think at worse req.body is undefined, which is why previously i didn't include it in try blocks. If you pass in an undefined argument to a mongoose call, and that throws an error, then technically the error is handled there.

File: server/api/categoryRouter.js (hunk 13 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:30]: discussed blow

File: server/test/categoryRouter.test.js (hunk 12 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js:51]: It can, but I only created it to pass as an argument. I think it is convention to be clear in the test that the err is expected to be null. That being said, feel free to change it if you like!

File: server/api/itemRouter.js (hunk 14 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:19]: we could handle these by doing if (newCategory) then proceed with the rest. Though i'm not sure that's necessary. Either works probably?

File: server/test/categoryRouter.test.js (hunk 12 above)
Comment (Sherry2001) [reply server/test/categoryRouter.test.js:51]: oh no it's okay, I agree. I see- so that helper was only for the mongoose callback.

File: server/api/itemRouter.js (hunk 14 above)
Comment (Anthony-Fiddes) [reply server/api/itemRouter.js:19]: I don't know enough to say. The server will currently respond with a 500 error, but I think 400 would be more appropriate (malformed request)

File: server/test/categoryRouter.test.js (hunk 12 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js:51]: yep!

File: server/index.js [hunk 15]
```diff
@@ -4,16 +4,27 @@ const cors = require('cors');
...
+}
 
-app.use(bodyParser.urlencoded({extended: true}));
+//Import Routers
+const categoryRouter = require('./api/categoryRouter.js');
+const itemRouter = require('./api/itemRouter.js');
+
+app.use(bodyParser.urlencoded({ extended: true }));
 app.use(cors());
 app.use(bodyParser.json());
```
Comment (Anthony-Fiddes) [inline server/index.js:20]: Line 18 and 20 are functionally the same. Also consider using express.json() and removing the extra import.

File: server/index.js [hunk 16]
```diff
@@ -4,16 +4,27 @@ const cors = require('cors');
 const app = express();
 const apiPort = 8080;
 
-const db = require('./db');
+//connected to mongoose through db/index.js
+const { getMongoDB, uri } = require('./db');
```
Comment (Anthony-Fiddes) [inline server/index.js:8]: This could actually go into the if statement as well

File: server/index.js (hunk 15 above)
Comment (Sherry2001) [reply server/index.js]: Yep, changed bodyParser to express, but left both for bc this was in the stack overflow link "You can use both if you want to. json is for JSON and urlencoded is for URL encoded data. If you add both you'll be able to parse both types." So i'll leave it for now in case we use it later?

File: server/index.js (hunk 16 above)
Comment (Sherry2001) [reply server/index.js:8]: ype

File: server/index.js (hunk 15 above)
Comment (Anthony-Fiddes) [reply server/index.js]: o, that's a very good reason! Yes let's keep it

File: server/api/itemRouter.js (hunk 14 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:19]: the error response would be from next(error), and the status code would depend on the error from the mongodb call's callback or whatever error the method catches. I don't think that is caused by req.body. 

If we want to change the error code in general, we can write a middleware error function to make it always give 400 as code

Comment (Anthony-Fiddes) [reply server/api/itemRouter.js:19]: I see, thank you!

File: server/db/index.js [hunk 17]
```diff
@@ -1,10 +1,19 @@
 const mongoose = require('mongoose');
 
-mongoose
-  .connect('mongodb+srv://user:password@lobstericecream-trflp.gcp.mongodb.net/home?retryWrites=true&w=majority')
-  .catch((e) => {
-    console.error('Connection error', e.message);
-  });
+const defaultUri = 'mongodb+srv://user:password@lobstericecream-trflp.gcp.mongodb.net/home?retryWrites=true&w=majority';
```
Comment (ricazhang) [inline server/db/index.js:3]: Ideally you don't want to store the password in Github. One option is to store it in a dotenv file.

File: server/api/itemRouter.js [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+ * response: {success: boolean,
+ *            message: String}
+ */
+router.put('/addItemToCategory', async (req, res, next) => {
+  let response = {}; 
+  try {
+    const itemId = req.body.itemId;
+    const categoryId = req.body.categoryId;
+    await Item.update({ _id: itemId }, { $push: { categoryIds: this.categoryId } }, done).exec();
+    await Category.update({ _id: categoryId }, { $push: { items: this.itemId } }, done).exec();
```
Comment (ricazhang) [inline server/api/itemRouter.js:72]: Would it be worth adding one more layer of verification by comparing the user ID of the item and category to the one in the request?

File: server/api/itemRouter.js [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    response.success = true;
+    response.message = 'item added to category';
```
Comment (ricazhang) [inline server/api/itemRouter.js:74]: This response message doesn't seem necessary unless you're planning on displaying it, in which case you should make it display friendly here.

File: server/api/itemRouter.js [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+ *            message: String}
+ */
+router.put('/addItemToCategory', async (req, res, next) => {
+  let response = {}; 
+  try {
+    const itemId = req.body.itemId;
+    const categoryId = req.body.categoryId;
+    await Item.update({ _id: itemId }, { $push: { categoryIds: this.categoryId } }, done).exec();
+    await Category.update({ _id: categoryId }, { $push: { items: this.itemId } }, done).exec();
+    response.success = true;
```
Comment (ricazhang) [inline server/api/itemRouter.js:73]: Since we aren't doing anything special with this response here, I think it'd be better to use the res "status" and "send" methods.

File: server/api/itemRouter.js [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    response.message = 'item added to category'; 
+    res.json(response);
+  } catch (error) {
+    response.success = false;
+    response.message = 'error adding item to category';
```
Comment (ricazhang) [inline server/api/itemRouter.js:78]: It might be helpful to clarify that this is caused by a db error as opposed to some other error. For example "Failed to update item/category in DB".

File: server/api/itemRouter.js [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+  let response = {};
+  try { 
+    const itemId = req.body.itemId; 
+    const itemObject = await Item.findOneAndRemove({ _id: itemId}).exec();
+    const itemCategoryIds = itemObject.categoryIds;
+
+    //deleting this itemId from all the categories it belonged to
+    await Promise.all(itemCategoryIds.map(async (categoryId) => {
+      await Category.update({ _id: categoryId}, { $pull: { items: this.itemId} }, done).exec();
+    }))
```
Comment (ricazhang) [inline server/api/itemRouter.js:102]: Not sure if this works but can you just use pull to go through all documents? I think it would look something like:
```
db.collection.update(
   { },
   { $pull: { items: this.itemId} },
  done).exec();
```
So that you don't have to send multiple requests?

File: server/api/itemRouter.js [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+ * req.body: {itemId: mongoose.objectId}
+ * 
+ * response: {success: boolean,
+ *            message: String}
+ */
+router.delete('/deleteItem', async (req, res, next) => {
+  let response = {};
+  try { 
+    const itemId = req.body.itemId; 
+    const itemObject = await Item.findOneAndRemove({ _id: itemId}).exec();
```
Comment (ricazhang) [inline server/api/itemRouter.js:96]: Are there any issues with removing the item before removing all references to it in the Categories?

File: server/test/categoryRouter.test.js [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    it('should return a JSON array with all categories associated with the given userId', async function () {
+      chai
+        .request(server)
+        .get('/category/getCategories')
+        .set('content-type', 'application/json')
+        .send({
+          userId: defaultUser._id
+        })
+        .end(function (err, res) {
+          expect(err).to.be.null;
```
Comment (ricazhang) [inline server/test/categoryRouter.test.js:66]: Can you add tests for when err is not null?

File: server/api/categoryRouter.js [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+    const categoryData =  await Category.findOne({ _id: categoryId}, 'title items').exec();
+    response.title = categoryData.title; 
+    const itemIds = categoryData.items;
+    const itemObjects = [];
+
+    //for each item id, actually get the item object
+    await Promise.all(itemIds.map(async (itemId) => {
+      const itemObject = await Item.findById(itemId).exec();
+      itemObjects.push(itemObject);
+    }));
```
Comment (ricazhang) [inline server/api/categoryRouter.js:75]: Does `find` work to just send one request?

File: server/api/categoryRouter.js [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+ *            success: boolean,
+ *            message: String,
+ *           }
+ * error 
+ */
+router.get('/deleteCategory', async (req, res, next) => {
+  try {
+    const categoryId = req.body.categoryId;
+    let response = {};
+    const category = await Category.find({ _id: categoryId }).exec();
```
Comment (ricazhang) [inline server/api/categoryRouter.js:98]: findOne?

File: server/api/itemRouter.js [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+ *            comment: String (optional),}
+ * 
+ * response: {success: boolean,
+ *            message: String}
+ */
+router.post('/addItem', (req, res, next) => {
+  const newItem = req.body;
+  const item = new Item(newItem);
+  let response = {}; 
+  item.save((err) => {
```
Comment (ricazhang) [inline server/api/itemRouter.js:22]: Unless there was a reason for this, can we enforce consistency between when to use await vs when to use a callback?

File: server/api/categoryRouter.js [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+ * response: [{title: String, 
+ *             items: [{
+ *               item objects from db 
+ *             }]
+ *            }]
+ */
+router.get('/getCategoryItems', async (req, res, next) => {
+  try {
+    const categoryId = req.body.categoryId;
+    let response = {};
```
Comment (ricazhang) [inline server/api/categoryRouter.js:64]: const if it isn't being reassigned

File: server/api/categoryRouter.js [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+ * 
+ * req.body: {categoryId: mongoose.objectId}
+ * 
+ * response: [{title: String, 
+ *             items: [{
+ *               item objects from db 
+ *             }]
+ *            }]
+ */
+router.get('/getCategoryItems', async (req, res, next) => {
```
Comment (ricazhang) [inline server/api/categoryRouter.js:61]: nit: avoid abbreviations/acronyms for variable/param names

File: server/test/categoryRouter.test.js [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+          expect(res).to.have.status(200);
+          expect(res.body).to.have.lengthOf(1);
```
Comment (ricazhang) [inline server/test/categoryRouter.test.js:68]: that.is.empty();

File: server/test/categoryRouter.test.js [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+const chaiHttp = require('chai-http');
+const { getMongoDB } = require('../db');
+const { MongoMemoryServer } = require('mongodb-memory-server');
+const mongoose = require('mongoose');
+const User = require('../db/models/user');
+
+let exampleCategory;
+let defaultUser;
+let mongoServer;
+const expect = chai.expect;
```
Comment (ricazhang) [inline server/test/categoryRouter.test.js:13]: Curious, why expect over assert?

File: server/api/categoryRouter.js [hunk 32]
```diff
@@ -0,0 +1,118 @@
...
+    await Category.deleteOne({ _id: categoryId }).exec();
+
+    this.response.success = true;
+    this.response.message = 'deleted category';
+    res.json(response);
+  } catch (error) {
+    this.response.success = false;
+    this.response.message = 'error when deleting';
+    res.json(response);
+    next(error);
```
Comment (Alfredhgoog) [inline server/api/categoryRouter.js:114]: Like above, you only need to 1) handle error by yourself by res.json(responseIndicatingError) or 2) let the default error handler handle it by passing next(err), but not both.

File: server/api/itemRouter.js (hunk 27 above)
Comment (Alfredhgoog) [reply server/api/itemRouter.js:22]: +1

File: server/api/itemRouter.js [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+    if (err) {
+      response.success = false;
+      response.message = 'error adding item';
+      res.json(response);
+      next(err);
```
Comment (Alfredhgoog) [inline server/api/itemRouter.js:27]: Same above doing either, not both

File: server/api/itemRouter.js [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+});
+
+/**
+ * Get all items
+ * 
+ * req.body: {userId: mongoose.objectId}
+ * 
+ * response: [{sourceLink: String,
+ *            highlight: String,
+ *            placeId: id(optional),
```
Comment (Alfredhgoog) [inline server/api/itemRouter.js:43]: ID is also a string type? To specify a field as optional, the goog closure style goes like
placeId: string=

But no strict requirement of adopting the whole documentation style guide.

File: server/api/itemRouter.js [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+ *            comments: String (optional),}]
+ */
+router.get('/getItems', (req, res, next) => {
+  Item.find({ userId: req.body.userId }, (err, items) => {
+    if (err) {
+      next(err);
+    }
+    if (items) {
+      res.json(items);
+    }
```
Comment (Alfredhgoog) [inline server/api/itemRouter.js:53]: If there's no 'items' I think this middleware will result in a broken pipe. But I think the if isn't necessary as a whole so you can just always res.json(items).

File: server/api/itemRouter.js (hunk 18 above)
Comment (Alfredhgoog) [reply server/api/itemRouter.js:72]: I agree this is safe, but verification seems to fit in a top level common middleware, applicable for a few handlers. Let's put a TODO and open an issue on this.

File: server/api/itemRouter.js [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+    await Item.update({ _id: itemId }, { $push: { categoryIds: this.categoryId } }, done).exec();
+    await Category.update({ _id: categoryId }, { $push: { items: this.itemId } }, done).exec();
+    response.success = true;
+    response.message = 'item added to category'; 
+    res.json(response);
+  } catch (error) {
+    response.success = false;
+    response.message = 'error adding item to category'; 
+    res.json(response);
+    next(error);
```
Comment (Alfredhgoog) [inline server/api/itemRouter.js:80]: Same on doing either.

File: server/index.js [hunk 37]
```diff
@@ -1,19 +1,29 @@
...
+  const mongoDB = getMongoDB(uri);
+  mongoDB.on('error', console.error.bind(console, 'MongoDB connection error: '));
+}
 
-app.use(bodyParser.urlencoded({extended: true}));
-app.use(cors());
-app.use(bodyParser.json());
+//Import Routers
+const categoryRouter = require('./api/categoryRouter.js');
+const itemRouter = require('./api/itemRouter.js');
```
Comment (Alfredhgoog) [inline server/index.js:15]: Please move require statements to the start of the file.

File: server/test/categoryRouter.test.js [hunk 38]
```diff
@@ -0,0 +1,74 @@
...
+after(async function () {
+  await mongoose.disconnect();
+  await mongoServer.stop();
+});
+
+it('There should be one user in the database', async function () {
+  await Category.find(function (err, categories) {
+    expect(err).to.be.null;
+    console.log(categories);
+  });
```
Comment (Alfredhgoog) [inline server/test/categoryRouter.test.js:39]: Is this a mix of the await and callback style? Please move the expect statements below the await call.

File: server/test/categoryRouter.test.js [hunk 39]
```diff
@@ -0,0 +1,74 @@
...
+describe('categoryRouter', function () {
+  describe('/getCategories', function () {
+    it('should return an empty list when no userId is provided in the request', async function () {
+      chai
+        .request(server)
+        .get('/category/getCategories')
+        .end(function (err, res) {
+          expect(err).to.be.null;
+          expect(res).to.have.status(200);
+          expect(res.body).to.have.lengthOf(0);
```
Comment (Alfredhgoog) [inline server/test/categoryRouter.test.js:53]: simplify as expect(res.body).to.be.empty;

File: server/test/categoryRouter.test.js (hunk 31 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js:13]: No particular reason! It was probably because all the tutorials I saw were using expect. We can change it if you like

File: server/api/categoryRouter.js [hunk 40]
```diff
@@ -0,0 +1,118 @@
...
+ * Get all of the user's categories- just names and ids.
+ * 
+ * req.body: {userId: mongoose.objectId}
+ * 
+ * response: [{ title: String, _id: Mongoose.objectId}]
+ */
+router.get('/getCategories', async (req, res, next) => {
+  try {
+    const userId = req.body.userId;
+    const categoriesData = await Category.find({ userId: userId }, 'title').exec();
```
[the remark this answers is not in the document]
Comment (Anthony-Fiddes) [reply server/api/categoryRouter.js]: That's really cool!

File: server/test/categoryRouter.test.js (hunk 30 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js:68]: I'm sorry, could you explain this comment more? Does it apply to err?

File: server/db/index.js (hunk 17 above)
Comment (Anthony-Fiddes) [reply server/db/index.js]: Added the dotenv solution

**Comment by Anthony-Fiddes** [comment]
Screenshot of the router being tested

image

**Comment by Sherry2001** [comment]
@Alfredhgoog @ricazhang Yes... Very sorry about this giant PR. We made the mistake of trying to combine our work with each other's first before presenting to you guys, which ended up getting very large. But I understand what you mean and will do that going forward. Thank you for going through this long one.

File: server/api/categoryRouter.js [hunk 41]
```diff
@@ -0,0 +1,118 @@
...
+ * req.body: {title: String}
+ * 
+ * response: {success: boolean, message: String}
+ */
+router.post('/createCategory', (req, res, next) => {
+  const newCategory = req.body;
+  let response = {};
+  Category.create(newCategory, (err, data) => {
+    if (err) {
+      response.successs = false;
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:21]: Fixed

File: server/api/categoryRouter.js [hunk 42]
```diff
@@ -0,0 +1,118 @@
...
+      response.message = 'failed to add category';
+      next(err);
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:23]: This next(err) uses the default error handler in express, or at least that was my impression after reading this [https://expressjs.com/en/guide/error-handling.html](url). But we were discussing potentially writing our own error handler.

File: server/api/itemRouter.js (hunk 18 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:72]: @ricazhang Oh.. I see what you mean. 

Is this check necessary considering all itemIds and categoryIds should be completely unique, so it would be unlikely that we get the wrong item or category belonging to a different user unless the request coming from the client-side is wrong, which could be possible but just our own mistake? Or is it better practice to do this check? 

I added TODO and opened an issue on this @Alfredhgoog

File: server/api/itemRouter.js (hunk 19 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:74]: Ah yeah...
Is it appropriate/standard to display a message in the console once a request was successfully complete or would that be considered a debugging message and should not happen?

File: server/api/itemRouter.js (hunk 21 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:78]: okay! fixed for all such error messages.

File: server/api/itemRouter.js (hunk 22 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:102]: Oh yes thank you! That is much smarter. Made this change for deleteItem and deleteCategory

File: server/api/itemRouter.js (hunk 23 above)
Comment (Sherry2001) [reply server/api/itemRouter.js]: Not sure, but will check when we test these. I changed the order for now to be safe.

File: server/api/itemRouter.js (hunk 19 above)
Comment (ricazhang) [reply server/api/itemRouter.js]: Yes console would be debugging message :)

File: server/api/categoryRouter.js (hunk 26 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js]: yes!

File: server/test/categoryRouter.test.js (hunk 30 above)
Comment (ricazhang) [reply server/test/categoryRouter.test.js:68]: Oops I left this comment on the wrong line. Meant to leave it at line 53 but it looks like Alfred already left a comment there in the meantime.

File: server/api/categoryRouter.js (hunk 25 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:75]: Yes thank you! Changed

File: server/api/categoryRouter.js (hunk 28 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:64]: okay!

File: server/api/categoryRouter.js (hunk 29 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:61]: Just to make sure- so would even the 'req' and 'res' not be abbreviated but changed into 'request' and 'response'?

File: server/api/itemRouter.js (hunk 27 above)
Comment (Sherry2001) [reply server/api/itemRouter.js]: Yes, fixed for all functions that still use callback.

File: server/api/categoryRouter.js [hunk 43]
```diff
@@ -0,0 +1,118 @@
...
+ * Create a new category
+ * 
+ * req.body: {title: String}
+ * 
+ * response: {success: boolean, message: String}
+ */
+router.post('/createCategory', (req, res, next) => {
+  const newCategory = req.body;
+  let response = {};
+  Category.create(newCategory, (err, data) => {
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js]: Yes, fixed these for consistency

File: server/api/itemRouter.js (hunk 35 above)
Comment (Sherry2001) [reply server/api/itemRouter.js]: ohh i see. Got it, and also fixed this to use await

File: server/api/categoryRouter.js [hunk 44]
```diff
@@ -0,0 +1,118 @@
...
+    if (err) {
+      response.successs = false;
+      response.message = 'failed to add category';
+      next(err);
+    } else {
+      response.success = true;
+      response.message = 'created new category';
+    }
+    res.json(response);
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:28]: I will use only next(err)

File: server/test/categoryRouter.test.js (hunk 24 above)
Comment (Sherry2001) [reply server/test/categoryRouter.test.js:66]: TODO for next PR

File: server/api/itemRouter.js (hunk 20 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:70]: okay will correct!

File: server/api/itemRouter.js (hunk 33 above)
Comment (Sherry2001) [reply server/api/itemRouter.js]: corrected

File: server/api/categoryRouter.js (hunk 32 above)
Comment (Sherry2001) [reply server/api/categoryRouter.js:105]: Choosing next(err). Corrected

File: server/test/categoryRouter.test.js (hunk 38 above)
Comment (Anthony-Fiddes) [reply server/test/categoryRouter.test.js]: Sorry about that! I didn't have a good understanding of the two styles

File: server/api/categoryRouter.js [hunk 45]
```diff
@@ -0,0 +1,118 @@
...
+});
+
+/**
+ * Get all of the user's categories- just names and ids.
+ * 
+ * req.body: {userId: mongoose.objectId}
+ * 
+ * response: [{ title: String, _id: Mongoose.objectId}]
+ */
+router.get('/getCategories', async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:39]: Thank you for the info. Got it- I switched get request to using req.params

File: server/api/categoryRouter.js [hunk 46]
```diff
@@ -0,0 +1,118 @@
...
+ * 
+ * req.body : {categoryId: mongoose.objectId}
+ * 
+ * response: {
+ *            success: boolean,
+ *            message: String,
+ *           }
+ * error 
+ */
+router.get('/deleteCategory', async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:90]: This was a mistake. I changed it to router.delete

File: server/api/itemRouter.js (hunk 34 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:41]: oh I see. Thank you, I'll use that style for this at least.

File: server/api/categoryRouter.js [hunk 47]
```diff
@@ -0,0 +1,118 @@
...
+});
+
+/**
+ * Get all the items belonging in a specified category
+ * 
+ * req.body: {categoryId: mongoose.objectId}
+ * 
+ * response: [{title: String, 
+ *             items: [{
+ *               item objects from db
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:57]: okay, fixed!

File: server/api/itemRouter.js (hunk 36 above)
Comment (Sherry2001) [reply server/api/itemRouter.js:77]: fixed to just next(error)

File: server/api/categoryRouter.js [hunk 48]
```diff
@@ -0,0 +1,118 @@
...
+    res.json(response);
+  });
+});
+
+/**
+ * Get all of the user's categories- just names and ids.
+ * 
+ * req.body: {userId: mongoose.objectId}
+ * 
+ * response: [{ title: String, _id: Mongoose.objectId}]
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply server/api/categoryRouter.js:37]: Okay, fixed

**Comment by Sherry2001** [comment]
@ricazhang  @Alfredhgoog  
Main Changes from my side for your comments: 

1) Sending one collective find request to the DB that gets the job done instead of numerous singular requests using Promise.all and map: c55e426 42a51ea 

2) Ensuring consistency in using async/await and try/catch blocks instead of callback functions; other smaller changes suggested by Rica: af0d2ca d057487

3) Consistency in using next(err) and res.status(200).send(message) instead of self-made response objects for success/error; documentation comment corrections; other smaller changes suggested by Alfred: 03d2056

**Comment by ricazhang** [comment]
Don't forget to run a formatter again!

**Comment by ricazhang** [comment]
Approval pending the requested changes

File: .gitignore [hunk 49]
```diff
@@ -1,2 +1,3 @@
 server/db/index.js
-node_modules
\ No newline at end of file
+node_modules
+.env
```
[the remark this answers is not in the document]
Comment (Sherry2001) [reply .gitignore]: done

File: server/api/itemRouter.js (hunk 19 above)
Comment (Sherry2001) [reply server/api/itemRouter.js]: okay, will explore a logging library later!

**Comment by Sherry2001** [comment]
Ready to be re-reviewed by Alfred, addressed all comments from Rica.
